components/eam/src/physics/cam/check_energy.F90

```diff
@@ -1308,6 +1111,125 @@ subroutine check_prect(state, tend, name, nstep, ztodt, prect)
+
+    if (icldliq > 1  .and.  icldice > 1 .and. irain > 1 .and. isnow > 1) then
+       do i = 1, ncol
+          call energy_helper_eam_def_column(u(i,:),v(i,:),T(i,:),q(i,1:pver,1:pcnst),&
+                                   ps(i),pdel(i,:),phis(i), &
+                                   ke(i),se(i),wv(i),wl(i),wi(i),wr(i),ws(i),te(i),tw(i) )                             
+       enddo
+    else
+       call endrun('energy_helper...column is not implemented if water forms do not exist')
+    endif
+
+  end subroutine energy_helper_eam_def
+
+  subroutine energy_helper_eam_def_column(u,v,T,q,ps,pdel,phis, &
+                                   ke,se,wv,wl,wi,wr,ws,te,tw, &
+                                   teloc,psterm)
+
+!state vars are of size psetcols,pver, so, not exactly correct
+    real(r8), intent(in) :: u(pver)
+    real(r8), intent(in) :: v(pver)
+    real(r8), intent(in) :: T(pver)
+    real(r8), intent(in) :: q(pver,pcnst)
+    real(r8), intent(in) :: ps
+    real(r8), intent(in) :: pdel(pver)
+    real(r8), intent(in) :: phis
+
+    real(r8), intent(inout) :: ke     ! vertical integral of kinetic energy
+    real(r8), intent(inout) :: se     ! vertical integral of static energy
+    real(r8), intent(inout) :: wv     ! vertical integral of water (vapor)
+    real(r8), intent(inout) :: wl     ! vertical integral of water (liquid)
+    real(r8), intent(inout) :: wi     ! vertical integral of water (ice)
+    real(r8), intent(inout) :: te     ! vertical integral of total energy
+    real(r8), intent(inout) :: tw     ! vertical integral of total water
+    real(r8), intent(inout) :: wr     ! vertical integral of rain
+    real(r8), intent(inout) :: ws     ! vertical integral of snow
+
+    real(r8), intent(inout), optional :: teloc(pver)
+    real(r8), intent(inout), optional :: psterm
+
+    integer :: i,k
+
+    ! Compute vertical integrals of dry static energy and water (vapor, liquid, ice)
+    ke = 0._r8
+    se = 0._r8
+    wv = 0._r8
+    wl = 0._r8
+    wi = 0._r8
+    wr = 0._r8
+    ws = 0._r8
+
+    !keep it bfb and fast
+    if (present(teloc) .and. present(psterm))then
+       teloc = 0.0; psterm = 0.0
+       do k = 1, pver
+          teloc(k) = 0.5_r8*(u(k)**2 + v(k)**2)*pdel(k)/gravit &
+                   + t(k)*cpair*pdel(k)/gravit &
+                   + (latvap+latice)*q(k,1       )*pdel(k)/gravit
+          teloc(k) = teloc(k) &
+                   + latice*(q(k,icldliq) + q(k,irain))*pdel(k)/gravit
+       end do
+       psterm = phis*ps/gravit
+    endif
+
+    do k = 1, pver
+       ke = ke + 0.5_r8*(u(k)**2 + v(k)**2)*pdel(k)/gravit
+       se = se +         t(k)*cpair*pdel(k)/gravit
+       wv = wv + q(k,1      )*pdel(k)/gravit
+    end do
+    se = se + phis*ps/gravit
+
+    do k = 1, pver
+       wl = wl + q(k,icldliq)*pdel(k)/gravit
+       wi = wi + q(k,icldice)*pdel(k)/gravit
+    end do
+
+    do k = 1, pver
+       wr = wr + q(k,irain)*pdel(k)/gravit
+       ws = ws + q(k,isnow)*pdel(k)/gravit
```

Comment:
We probably need to add a conditional here... I'll add the change.

Reply:
Thanks!